mlir/lib/Dialect/Rock/Transforms/RemoveRedundantCasts.cpp

```diff
@@ -0,0 +1,510 @@
+      // that will potentially be redundant.
+      changed |= handleRockGemmWrapper(input, generic, rewriter);
+    } else if (auto inputGeneric = dyn_cast<linalg::GenericOp>(input)) {
+      // We also want to handle an arbitrary trunc op followed by ext ops
+      if (isGenericWithSingleOp<arith::TruncFOp>(inputGeneric) ||
+          isGenericWithSingleOp<arith::TruncIOp>(inputGeneric)) {
+        changed |= handleLinalgGeneric(inputGeneric, generic, rewriter);
+      }
+    } else {
+      // We have come across a pattern that we do not know how to handle, or
+      // does not require changing.
+      return failure();
+    }
+
+    return changed ? success() : failure();
+  }
+
+private:
+  // This is a helper function that checks if the generic operation contains
+  // only a single operation of type OpType, and that the yield returns that
+  // value.
+  // NOTE: A convert in migraphx, will eventually be lowered to a linalg
+  // with either a single trunc or ext op. e.g.,:
+  // %1 = migraphx.convert %0 : <1x5x3xf16, 15x3x1> to <1x5x3xf32, 15x3x1>
+  template <typename OpType>
+  bool isGenericWithSingleOp(linalg::GenericOp generic) const {
+    Block &body = generic.getRegion().front();
+
+    // Check that the body has exactly 2 operations: OpType and yield
+    if (std::distance(body.begin(), body.end()) != 2) {
+      return false;
+    }
+
+    // First operation should be an OpType
+    auto firstOp = body.begin();
+    if (!isa<OpType>(firstOp)) {
+      return false;
+    }
+    auto requiredOp = cast<OpType>(firstOp);
+
+    // Second operation should be yield
+    auto secondOp = std::next(body.begin());
+    if (!isa<linalg::YieldOp>(secondOp)) {
+      return false;
+    }
+    auto yieldOp = cast<linalg::YieldOp>(secondOp);
+
+    // Verify that the yield operation yields the result of the required op
+    return yieldOp.getValues().size() == 1 &&
+           yieldOp.getValues()[0] == requiredOp.getResult();
+  }
+
+  Value getExtInput(linalg::GenericOp generic) const {
+    // Check that there is only one input to the generic operation
+    if (generic.getInputs().size() != 1) {
+      LLVM_DEBUG(llvm::dbgs()
+                 << "\tGeneric does not have exactly one input, skipping\n");
+      return nullptr;
+    }
+
+    // We don't need to investigate BlockArguments any further
+    Value input = generic.getInputs()[0];
+    if (isa<BlockArgument>(input))
+      return nullptr;
+
+    // If the input is a rock.transform op, we need to untransform it to get
+    // to the original input value.
+    if (auto rockTransformOp =
+            dyn_cast<rock::TransformOp>(input.getDefiningOp())) {
+      // We need to traverse the rock.transform ops to find the original input
+      SmallVector<rock::TransformOp> inputTransforms;
+      auto untransformed = std::get<0>(untransform(input, inputTransforms));
+      if (isa<BlockArgument>(untransformed)) {
+        return nullptr;
+      }
+      input = untransformed;
```

Comment:
You can use `rock::untransform` directly without firsthcekcing it's defining op is rock::TransformOp or not.